File: ejemplos/general/05-generador-de-contrasenas/ejemplo_secrets.py
"""
Ejemplo: Construir un PIN bancario seguro con secrets + string

Qué enseña este ejemplo:
  - Cuándo usar secrets en lugar de random (y por qué importa)
  - Cómo usar string.digits, string.ascii_letters, string.punctuation
  - Cómo construir una cadena aleatoria con join() + secrets.choice()
  - Cómo garantizar que el resultado incluya al menos un carácter de cada tipo

Para tu reto (generador de contraseñas):
  Aplica el mismo patrón: construye el pool de caracteres según los
  parámetros del usuario (longitud, símbolos sí/no, mayúsculas sí/no),
  y usa secrets.choice() en un loop para construir la contraseña.

Glosario de términos "raros":
  secrets          — módulo para aleatoriedad segura (usa fuentes del sistema).
  random           — módulo de aleatoriedad predecible (NO usar para contraseñas).
  string.digits    — contiene "0123456789".
  string.ascii...  — contiene letras ("abcdefghijklmnopqrstuvwxyz", etc).
  string.punctuation — todos los símbolos: !"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~
  "".join(lista)   — une una lista de caracteres en un solo texto (ej: "a" + "b" = "ab").
  _ (guion bajo)   — se usa en `for _ in range()` cuando vas a repetir algo N veces 
                     pero no necesitas usar el número de la vuelta actual.
"""

import secrets
import string


def generar_pin(longitud: int = 6, solo_numeros: bool = True) -> str:
    """
    Genera un PIN seguro.
    Si solo_numeros=True, usa únicamente dígitos (PIN bancario clásico).
    Si solo_numeros=False, mezcla letras y dígitos (PIN alfanumérico).
    """
    if solo_numeros:
        pool = string.digits  # "0123456789"
    else:
        pool = string.ascii_uppercase + string.digits

    # join() + list comprehension: llama secrets.choice(pool) N veces
    # y une todos los caracteres resultantes en un string
    pin = "".join(secrets.choice(pool) for _ in range(longitud))
    return pin


def generar_pin_garantizado(longitud: int = 8) -> str:
    """
    Genera un PIN que GARANTIZA tener al menos:
      - 1 letra mayúscula
      - 1 dígito
      - 1 símbolo

    Técnica: generar una de cada tipo obligatorio, luego rellenar
    el resto con caracteres del pool completo y mezclar.
    """
    if longitud < 3:
        raise ValueError("La longitud mínima para un PIN garantizado es 3.")

    pool_completo = string.ascii_uppercase + string.digits + string.punctuation

    # 1. Garantizar al menos uno de cada tipo requerido
    obligatorios = [
        secrets.choice(string.ascii_uppercase),
        secrets.choice(string.digits),
        secrets.choice(string.punctuation),
    ]

    # 2. Rellenar el resto con caracteres aleatorios del pool completo
    relleno = [secrets.choice(pool_completo) for _ in range(longitud - 3)]

    # 3. Unir todo y mezclar (secrets.SystemRandom para barajar de forma segura)
    caracteres = obligatorios + relleno
    secrets.SystemRandom().shuffle(caracteres)

    return "".join(caracteres)


if __name__ == "__main__":
    print("=== Generador de PINs seguros ===\n")

    print("PINs numéricos (6 dígitos):")
    for _ in range(3):
        print(f"  {generar_pin(6, solo_numeros=True)}")

    print("\nPINs alfanuméricos (8 caracteres):")
    for _ in range(3):
        print(f"  {generar_pin(8, solo_numeros=False)}")

    print("\nPINs con garantía (letra + dígito + símbolo, 10 chars):")
    for _ in range(3):
        print(f"  {generar_pin_garantizado(10)}")

File: archivo-historico/ejemplos/general/05-generador-de-contrasenas/ejemplo_secrets.rs
/*
Ejemplo: Construir un PIN bancario seguro con generadores de números aleatorios
 
Qué enseña este ejemplo:
  - Cómo usar el crate 'rand' para generar aleatoriedad criptográficamente segura (rand usa el OS RNG por defecto en thread_rng())
  - Cómo estructurar strings de constantes y mezclarlas
  - Cómo garantizar que el resultado incluya al menos un carácter de cada tipo
 
Para tu reto (generador de contraseñas):
  Aplica el mismo patrón: construye el pool de caracteres según los
  parámetros del usuario (longitud, símbolos sí/no, mayúsculas sí/no),
  y usa el RNG para construir la contraseña.
 
Glosario:
  rand::thread_rng() — obtiene el generador de números aleatorios seguro local
  slice::choose()    — selecciona un elemento aleatorio de una lista
  slice::shuffle()   — mezcla los elementos de una lista
*/

use rand::seq::SliceRandom;

const DIGITOS: &[u8] = b"0123456789";
const MAYUSCULAS: &[u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZ";
const SIMBOLOS: &[u8] = b"!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~";

fn generar_pin(longitud: usize, solo_numeros: bool) -> String {
    let mut rng = rand::thread_rng();
    
    let mut pool = Vec::new();
    if solo_numeros {
        pool.extend_from_slice(DIGITOS);
    } else {
        pool.extend_from_slice(MAYUSCULAS);
        pool.extend_from_slice(DIGITOS);
    }

    let mut pin = String::with_capacity(longitud);
    for _ in 0..longitud {
        let ch = pool.choose(&mut rng).unwrap();
        pin.push(*ch as char);
    }
    pin
}

fn generar_pin_garantizado(longitud: usize) -> String {
    if longitud < 3 {
        panic!("La longitud mínima para un PIN garantizado es 3.");
    }

    let mut rng = rand::thread_rng();
    
    let mut pool_completo = Vec::new();
    pool_completo.extend_from_slice(MAYUSCULAS);
    pool_completo.extend_from_slice(DIGITOS);
    pool_completo.extend_from_slice(SIMBOLOS);

    let mut caracteres = Vec::new();
    
    // 1. Garantizar
    caracteres.push(*MAYUSCULAS.choose(&mut rng).unwrap());
    caracteres.push(*DIGITOS.choose(&mut rng).unwrap());
    caracteres.push(*SIMBOLOS.choose(&mut rng).unwrap());

    // 2. Rellenar
    for _ in 3..longitud {
        caracteres.push(*pool_completo.choose(&mut rng).unwrap());
    }

    // 3. Mezclar
    caracteres.shuffle(&mut rng);

    // Convertir de bytes a String
    caracteres.iter().map(|&b| *b as char).collect()
}

fn main() {
    println!("=== Generador de PINs seguros ===\n");

    println!("PINs numéricos (6 dígitos):");
    for _ in 0..3 {
        println!("  {}", generar_pin(6, true));
    }

    println!("\nPINs alfanuméricos (8 caracteres):");
    for _ in 0..3 {
        println!("  {}", generar_pin(8, false));
    }

    println!("\nPINs con garantía (letra + dígito + símbolo, 10 chars):");
    for _ in 0..3 {
        println!("  {}", generar_pin_garantizado(10));
    }
}
